Admin X activitypub app independent release

allouis opened the pull request:
ref https://linear.app/tryghost/issue/AP-438

We want to have continuous deployment (or thereabouts) for the activitypub frontend, and to do that we're going to load it from the jsdelivr cdn, and deploy via npm package updates.

allouis committed:
Removed private flag

This is necessary so that we can publish to npm

allouis committed:
Allowed to load AdminX apps from custom url

ref https://linear.app/tryghost/issue/AP-438

This is going to allow us to load the activitypub package from the jsdelivr
cdn, which means we can release new versions without releasing the admin.

allouis committed:
Loaded AdminX activitypub app from jsdelivr in production

ref https://linear.app/tryghost/issue/AP-438

We want to keep development and test mode using the local build so we can have
feedback on our local changes.

kevinansfield commented on ghost/admin/lib/asset-delivery/index.js:
```diff
@@ -43,6 +43,8 @@ module.exports = {
                     console.log(`Asset-Delivery: ${key} = ${value}`);
                 }
+
+                this.packageConfig[`adminXActivitypubCustomUrl`] = 'https://cdn.jsdelivr.net/npm/@tryghost/admin-x-activitypub@0.0.2/dist/admin-x-activitypub.js'
```
Normal practice is to [set it Ghost's default config](https://github.com/TryGhost/Ghost/blob/main/ghost/core/core/shared/config/defaults.json#L183-L203) with a `minor` version listed so patches are automatically picked up immediately without having to ship Admin but you can still ship a minor to delay it being picked up before the next Ghost release.

What was the reason for trying to hardcode it in Admin? If you do that you'd need to rebuild and ship Ghost each time you want to bump the activity pub package.

kevinansfield replied:
> I tried added a new config property to config/environment.js but I get build errors, and I'm also not certain on where the environment variables are set for production builds.

What build error did you get? It should be possible to add a key to the `ENV` object without any issues 🤔

allouis replied:
RE: The version - that's the plan! This was just my WIP to get things working.

So does the Admin have access to the ghost config at build time? I didn't realise if that's the case!

The reason for hardcoding it was to just get things working and prove the concept

allouis committed:
Added workflow for deploying @tryghost/admin-x-activitypub

ref https://linear.app/tryghost/issue/AP-438

This will build and release the Admin X activitypub app when we bumb the
package.json version and push to `main`

allouis committed:
Updated @tryghost/admin-x-activitypub to 0.1.0

This will test the release when we merge to main

daniellockyer commented on .github/workflows/ci.yml:
```diff
@@ -965,3 +965,59 @@ jobs:
+        with:
+          node-version: '18.12.1'
+
+      - name: Restore caches
+        uses: actions/cache@v4
+        with:
+          path: |
+            ${{ github.workspace }}/apps/admin-x-activitypub/node_modules
+          key: admin-x-activitypub-${{ github.sha }}
+          restore-keys: |
+            admin-x-activitypub-
+
+      - name: Install dependencies
+        run: yarn install --prefer-offline --frozen-lockfile
```
You should be able to re-use our cache here: https://github.com/TryGhost/Ghost/blob/5579f03fd5fcc13c308f8c57c2e448dc10ad8e07/.github/workflows/ci.yml#L261-L264

It includes the `node_modules` for all packages but it's much faster than a `yarn install`

daniellockyer commented on .github/workflows/ci.yml:
```diff
@@ -965,3 +965,59 @@ jobs:
+        working-directory: apps/admin-x-activitypub
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
+        run: npm publish --dry-run --access public
```
One thing we might want to do after this is to auto-purge the jsDelivr cache

allouis replied:
Yeah, I was thinking this too

daniellockyer commented on ghost/admin/lib/asset-delivery/index.js:
```diff
@@ -43,6 +43,8 @@ module.exports = {
                     console.log(`Asset-Delivery: ${key} = ${value}`);
                 }
+
+                this.packageConfig[`adminXActivitypubCustomUrl`] = 'https://cdn.jsdelivr.net/npm/@tryghost/admin-x-activitypub@0.x/dist/admin-x-activitypub.js'
```
Does this work? 🤔 I thought it had to be `...@~0.0/..`

allouis replied:
It is working - the docs say it should be `@0` or `@0.0` though - I'm gonna change it to that
